### PR TITLE
Wrap non-pageable list functions into pageable

### PR DESCRIPTION
**Pull request by ArcturusZhang**
Fixes https://github.com/Azure/autorest.csharp/issues/1642

# Description

This PR wraps the non-pageable list result into pageable result to produce more consistent APIs in the mgmt plane SDKs.
This PR also makes some refactor to some existing code.

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

File: samples/Azure.ResourceManager.Sample/Generated/VirtualMachineScaleSetVirtualMachineExtensionCollection.cs [hunk 1]
```diff
@@ -293,20 +294,25 @@ public async virtual Task<Response<bool>> CheckIfExistsAsync(string vmExtensionN
...
+            Page<VirtualMachineScaleSetVirtualMachineExtension> FirstPageFunc(int? pageSizeHint)
             {
-                var response = _virtualMachineScaleSetVMExtensionsRestClient.List(Id.SubscriptionId, Id.ResourceGroupName, Id.Parent.Name, Id.Name, expand, cancellationToken);
-                return Response.FromValue(response.Value.Value.Select(value => new VirtualMachineScaleSetVirtualMachineExtension(Parent, value)).ToArray() as IReadOnlyList<VirtualMachineScaleSetVirtualMachineExtension>, response.GetRawResponse());
-            }
-            catch (Exception e)
-            {
-                scope.Failed(e);
-                throw;
+                using var scope = _clientDiagnostics.CreateScope("VirtualMachineScaleSetVirtualMachineExtensionCollection.GetAll");
```
Comment (AlexanderSher) [inline samples/Azure.ResourceManager.Sample/Generated/VirtualMachineScaleSetVirtualMachineExtensionCollection.cs:302]: This code is different from the previous one semantically. Previously, user would've got fully initialized collection after method execution. Now, `FirstPageFunc` will be called only during enumeration. Is that expected?

Comment (ArcturusZhang) [reply samples/Azure.ResourceManager.Sample/Generated/VirtualMachineScaleSetVirtualMachineExtensionCollection.cs:302]: Well, this is not expected. Do you have a suggestion how we could restore previous behavior? Does ``PageableHelpers` have related APIs to do this? Like I first call the API and get the array then I pass it into some function and it gives me a `Pageable` of something?

Comment (AlexanderSher) [reply samples/Azure.ResourceManager.Sample/Generated/VirtualMachineScaleSetVirtualMachineExtensionCollection.cs:302]: Sure. You can use `FromPages`:
https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/Pageable.cs#L97
https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/AsyncPageable.cs#L118
e.g.:
```cs
using var scope = _clientDiagnostics.CreateScope("VirtualMachineScaleSetVirtualMachineExtensionCollection.GetAll");
scope.Start();
try
{
    var response = _virtualMachineScaleSetVMExtensionsRestClient.List(Id.SubscriptionId, Id.ResourceGroupName, Id.Parent.Name, Id.Name, expand, cancellationToken: cancellationToken);
    var page = Page.FromValues(response.Value.Value.Select(value => new VirtualMachineScaleSetVirtualMachineExtension(Parent, value)), null, response.GetRawResponse());
    return Pageable<T>.FromPages(new[]{page});
}
catch (Exception e)
{
    scope.Failed(e);
    throw;
}
```

Comment (ArcturusZhang) [reply samples/Azure.ResourceManager.Sample/Generated/VirtualMachineScaleSetVirtualMachineExtensionCollection.cs:302]: Thank you! Let me switch to this and confirm with Michael

Comment (ArcturusZhang) [reply samples/Azure.ResourceManager.Sample/Generated/VirtualMachineScaleSetVirtualMachineExtensionCollection.cs:301]: I suppose this should be expected. We could merge with this behaviour and if this is not good enough, we could revisit this piece of logic. Changing this is not a breaking change for the SDK API.

**Pull request merged**